src/Core/DecimalFunctions.h

```diff
@@ -154,11 +155,20 @@ inline DecimalComponents<DecimalType> splitWithScaleMultiplier(
         typename DecimalType::NativeType scale_multiplier)
 {
     using T = typename DecimalType::NativeType;
-    const auto whole = decimal.value / scale_multiplier;
+    auto whole = decimal.value / scale_multiplier;
     auto fractional = decimal.value % scale_multiplier;
     if (fractional < T(0))
```

Comment:
>  When DecimalType is DateTime64, fractional is always positive, even if whole is negative.

could you pls clarify, why it is only about `DateTime64`? seems like this `if` should make `fractional` nonnegative in all cases?

Reply:
@nickitat See this issue: https://github.com/ClickHouse/ClickHouse/issues/34831
For datetime64 before `1970-01-01 00:00:00 UTC`, for example `1965-12-12 12:12:12.123`, its unit timestamp in ms unit is 
`-127943268000 + 123`, not `127943268000 - 123`

Reply:
IMO the main question is why do we get 127943268 as the `whole`. it seems like the input `decimal` has a wrong value, isn't it?
if you suspect that the pair `splitWithScaleMultiplier` + `decimalFromComponentsWithMultiplier` works incorrectly, you can try to reproduce it with [unit test](https://github.com/ClickHouse/ClickHouse/blob/master/src/Core/tests/gtest_DecimalFunctions.cpp).

Reply:
@nickitat Yes. The decimal value tranformed from DateTime64 before `1970-01-01 00:00:00 UTC` is wrong. That's why we have to fix these two functions `splitWithScaleMultiplier` and `decimalFromComponentsWithMultiplier`

Reply:
1. DateTime64(scale=3) -> Decimal
DateTime64: 1965-12-12 12:12:12.123 UTC
Expected value: -127943268000 + 123 = -127943267877
Real value:  -127943268000 - 123 = -127943268123

2. Decimal -> DateTime64(scale=3)
Decimal: -127943268000 + 123 = -127943267877  
Expected value:  whole = - 127943268 + 123 = 1965-12-12 12:12:12.123
Real value:  whole = -127943267  fractional = -877  -> 1965-12-12 12:12:13.877